lib/srv/db/redis/engine.go

```diff
@@ -0,0 +1,248 @@
+	}
+
+	if _, err := e.proxyConn.Write(buf.Bytes()); err != nil {
+		return trace.ConnectionProblem(err, "failed to send message to the client")
+	}
+
+	return nil
+}
+
+// HandleConnection is responsible for connecting to a Redis instance/cluster and
+func (e *Engine) HandleConnection(ctx context.Context, sessionCtx *common.Session) error {
+	// Check that the user has access to the database.
+	err := e.authorizeConnection(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	tlsConfig, err := e.Auth.GetTLSConfig(ctx, sessionCtx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	connectionOptions, err := ParseRedisURI(sessionCtx.Database.GetURI())
+	if err != nil {
+		return trace.BadParameter("Redis connection string is incorrect: %v", err)
+	}
+
+	var (
+		redisConn      redis.UniversalClient
+		connectionAddr = fmt.Sprintf("%s:%s", connectionOptions.address, connectionOptions.port)
+	)
+
+	// TODO(jakub): Use system CA bundle if connecting to AWS.
+	// TODO(jakub): Investigate Redis Sentinel.
+	switch connectionOptions.mode {
+	case Standalone:
+		redisConn = redis.NewClient(&redis.Options{
+			Addr:      connectionAddr,
+			TLSConfig: tlsConfig,
+		})
+	case Cluster:
+		if sessionCtx.DatabaseName != "" {
+			// ref: https://redis.io/commands/select
+			return trace.BadParameter("Redis Cluster only supports database zero")
+		}
+
+		redisConn = redis.NewClusterClient(&redis.ClusterOptions{
+			Addrs:     []string{connectionAddr},
+			TLSConfig: tlsConfig,
+		})
+	default:
+		// We've checked that while validating the config, but checking again can help with regression.
+		return trace.BadParameter("incorrect connection mode %s", connectionOptions.mode)
+	}
+
+	defer func() {
+		if err := redisConn.Close(); err != nil {
+			e.Log.Errorf("failed to close Redis connection: %v", err)
+		}
+	}()
+
+	e.Audit.OnSessionStart(e.Context, sessionCtx, nil)
+	defer e.Audit.OnSessionEnd(e.Context, sessionCtx)
+
+	if err := e.process(ctx, redisConn); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// readClientCmd reads commands from connected Redis client.
+func (e *Engine) readClientCmd(ctx context.Context) (*redis.Cmd, error) {
+	cmd := &redis.Cmd{}
+	if err := cmd.ReadReply(e.clientReader); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	val, ok := cmd.Val().([]interface{})
+	if !ok {
+		return nil, trace.BadParameter("failed to cast Redis value to a slice")
+	}
+
+	return redis.NewCmd(ctx, val...), nil
+}
+
+// process is the main processing function for Redis. It reads commands passed from client and passes them to
+// a Redis instance. It's also responsible for audit.
+func (e *Engine) process(ctx context.Context, redisClient redis.UniversalClient) error {
+	for {
+		// Read commands from client.
+		cmd, err := e.readClientCmd(ctx)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		// send valid commands to Redis instance/cluster.
+		err = e.processCmd(ctx, redisClient, cmd)
```

Comment:
Interesting question. I've checked that and looks like `QUIT` (at least in case of `redis-cli`) just closes client connection without sending anything to the server. In this case `processCmd` is not called. I'll check that with some other clients, but I'd assume similar behaviour.